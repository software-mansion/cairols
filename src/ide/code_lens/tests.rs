use std::ops::Not;

use super::{
    AnnotatedNode, FileCodeLens, LSCodeLens, LSCodeLensInterface, collect_functions_with_attrs,
    get_original_node_and_file, make_lens_args,
};
use crate::config::{Config, TestRunner};
use crate::lang::db::AnalysisDatabase;
use crate::lang::db::LsSemanticGroup;
use crate::lang::db::LsSyntaxGroup;
use crate::lang::lsp::ToCairo;
use crate::lang::lsp::{LsProtoGroup, ToLsp};
use crate::lsp::ext::ExecuteInTerminal;
use crate::lsp::ext::ExecuteInTerminalParams;
use crate::server::client::Notifier;
use crate::state::State;
use cairo_lang_defs::db::DefsGroup;
use cairo_lang_defs::ids::FreeFunctionLongId;
use cairo_lang_defs::ids::ModuleFileId;
use cairo_lang_defs::ids::ModuleId;
use cairo_lang_defs::ids::ModuleItemId;
use cairo_lang_defs::ids::SubmoduleLongId;
use cairo_lang_defs::ids::TopLevelLanguageElementId;
use cairo_lang_defs::plugin::MacroPlugin;
use cairo_lang_filesystem::db::FilesGroup;
use cairo_lang_filesystem::ids::CrateId;
use cairo_lang_syntax::node::helpers::QueryAttrs;
use cairo_lang_syntax::node::ids::SyntaxStablePtrId;
use cairo_lang_syntax::node::{TypedStablePtr, TypedSyntaxNode, ast::ModuleItem};
use cairo_lang_test_plugin::TestPlugin;
use cairo_lang_utils::Intern;
use lsp_types::Range;
use lsp_types::notification::ShowMessage;
use lsp_types::{CodeLens, Url};
use lsp_types::{Command, ShowMessageParams};
use crate::lsp::capabilities::client::ClientCapabilitiesExt;
use lsp_types::{MessageType, Position};
use serde_json::Number;
use serde_json::Value;
use super::CodeLensProvider;
use crate::config::Config;
use crate::config::TestRunner;
use crate::lang::db::AnalysisDatabase;
use crate::lang::db::LsSemanticGroup;
use crate::lang::db::LsSyntaxGroup;
use crate::lang::lsp::LsProtoGroup;
use crate::lang::lsp::ToCairo;
use crate::lang::lsp::ToLsp;
use crate::lsp::ext::ExecuteInTerminal;
use crate::lsp::ext::ExecuteInTerminalParams;
use crate::server::client::Notifier;
use crate::state::State;


#[derive(PartialEq, Clone, Debug)]
pub struct TestCodeLens {
    pub lens: CodeLens,
}

impl LSCodeLensInterface for TestCodeLens {
    fn execute(&self, file_url: Url, state: &State, notifier: &Notifier) -> Option<()> {
        let db = &state.db;

        let position = self.lens.range.start.to_cairo();

        let file = db.file_for_url(&file_url)?;
        let file_path = file_url.to_file_path().ok()?;

        let node = db.find_syntax_node_at_position(file, position)?;

        let module_item = node.ancestor_of_type::<ModuleItem>(db)?;
        let module_file_id = db.find_module_file_containing_node(module_item.as_syntax_node())?;

        let command = state.config.test_runner.command(
            TestFullQualifiedPath::new(db, module_item, module_file_id)?,
            AvailableTestRunners::new(db, module_file_id.0.owning_crate(db))?,
            &state.config.run_test_command,
        )?;

        let cwd = state.project_controller.configs_registry().manifest_dir_for_file(&file_path)?;
        if state.client_capabilities.execute_in_terminal_support() {
            notifier.notify::<ExecuteInTerminal>(ExecuteInTerminalParams { cwd, command });
        } else {
            notifier.notify::<ShowMessage>(ShowMessageParams {
                typ: MessageType::INFO,
                message: format!(
                    "To execute the code lens, run command: `{command}` in directory {}",
                    cwd.display()
                ),
            });
        }

        Some(())
    }

    fn get_lens(&self) -> CodeLens {
        self.lens.clone()
    }
}

pub fn push_test_code_lenses(
    file_code_lens: &mut FileCodeLens,
    url: Url,
    db: &AnalysisDatabase,
    config: &Config,
) -> Option<()> {
    let file = db.file_for_url(&url)?;

    let main_module = *db.file_modules(file).ok()?.first()?;

    let is_runner_available = config
        .test_runner
        .command(
            TestFullQualifiedPath::Function(String::new()), // We can substitute with anything here.
            AvailableTestRunners::new(db, main_module.owning_crate(db))?,
            &config.run_test_command,
        )
        .is_some();

    if is_runner_available {
        collect_test_lenses(db, main_module, url, file_code_lens);
    }

    Some(())
}

enum TestFullQualifiedPath {
    Function(String),
    Module(String),
}

impl TestFullQualifiedPath {
    fn cairo_test_command(&self) -> String {
        format!("scarb cairo-test --filter {}", self.as_ref())
    }

    fn snforge_command(&self) -> String {
        match self {
            TestFullQualifiedPath::Function(path) => {
                format!("snforge test {path} --exact")
            }
            TestFullQualifiedPath::Module(path) => {
                format!("snforge test {path}")
            }
        }
    }
}

impl AsRef<str> for TestFullQualifiedPath {
    fn as_ref(&self) -> &str {
        match self {
            TestFullQualifiedPath::Function(path) | TestFullQualifiedPath::Module(path) => path,
        }
    }
}

impl TestFullQualifiedPath {
    pub fn new(
        db: &AnalysisDatabase,
        module_item: ModuleItem,
        module_file_id: ModuleFileId,
    ) -> Option<Self> {
        match module_item {
            ModuleItem::FreeFunction(function_with_body) => {
                let path = ModuleItemId::FreeFunction(
                    FreeFunctionLongId(module_file_id, function_with_body.stable_ptr(db))
                        .intern(db),
                )
                .full_path(db);

                Some(TestFullQualifiedPath::Function(path))
            }
            ModuleItem::Module(item_module) => {
                let path = ModuleItemId::Submodule(
                    SubmoduleLongId(module_file_id, item_module.stable_ptr(db)).intern(db),
                )
                .full_path(db);

                Some(TestFullQualifiedPath::Module(path))
            }
            _ => None,
        }
    }
}

struct AvailableTestRunners {
    cairo_test: bool,
    snforge: bool,
}

impl AvailableTestRunners {
    pub fn new<'db>(db: &'db AnalysisDatabase, crate_id: CrateId<'db>) -> Option<Self> {
        let cairo_test = db.crate_macro_plugins(crate_id).iter().any(|plugin_id| {
            plugin_id.long(db).plugin_type_id() == TestPlugin::default().plugin_type_id()
        });

        // This will not work with crate renames in `Scarb.toml`, but there is no better way to do this now.
        let snforge = db.crate_config(crate_id)?.settings.dependencies.contains_key("snforge_std");

        Some(Self { cairo_test, snforge })
    }
}

impl TestRunner {
    fn command(
        &self,
        test_path: TestFullQualifiedPath,
        available_runners: AvailableTestRunners,
        custom_command: &str,
    ) -> Option<String> {
        match self {
            Self::Auto => match (available_runners.cairo_test, available_runners.snforge) {
                (true, false) => Some(test_path.cairo_test_command()),
                (false, true) => Some(test_path.snforge_command()),
                _ => None,
            },
            Self::CairoTest if available_runners.cairo_test => Some(test_path.cairo_test_command()),
            Self::Snforge if available_runners.snforge => Some(test_path.snforge_command()),
            Self::Custom => Some(custom_command.replace("{{TEST_PATH}}", test_path.as_ref())),
            _ => None,
        }
    }
}

const TEST_EXECUTABLES: [&str; 2] = ["test", "snforge_internal_test_executable"];
fn collect_test_functions(db: &AnalysisDatabase, module: ModuleId) -> Vec<AnnotatedNode> {
    collect_functions_with_attrs(db, module, &TEST_EXECUTABLES)
}

fn collect_tests<'db>(
    db: &'db AnalysisDatabase,
    module: ModuleId<'db>,
    file_url: Url,
    file_state: &mut FileCodeLens,
) {
    for node in collect_test_functions(db, module) {
        let index = file_state.len();
        maybe_push_code_lens(
            db,
            file_state,
            |position| make_test_code_lens(&file_url, position, false, index),
            node,
        );
    }

    let Ok(modules) = db.module_submodules_ids(module) else { return };

    for submodule in modules.iter().copied() {
        let is_inline = db.is_submodule_inline(submodule);

        let has_tests = if is_inline {
            let tests_count = file_state.len();

            collect_test_lenses(db, ModuleId::Submodule(submodule), file_url.clone(), file_state);

            // Append mod only if it contains tests.
            tests_count != file_state.len()
        } else {
            has_any_test(db, ModuleId::Submodule(submodule))
        };

        if has_tests {
            let ptr = submodule.stable_ptr(db).untyped();

            let full_path = submodule.full_path(db);
            let module_node = AnnotatedNode { full_path: full_path.clone(), attribute_ptr: ptr };
            let index = file_state.len();
            maybe_push_code_lens(
                db,
                file_state,
                |position| make_test_code_lens(&file_url, position, true, index),
                module_node,
            );
        }
    }
}

fn has_any_test<'db>(db: &'db AnalysisDatabase, module: ModuleId<'db>) -> bool {
    if collect_test_functions(db, module).is_empty().not() {
        return true;
    }

    let Ok(modules) = db.module_submodules_ids(module) else { return false };

    modules.iter().copied().map(ModuleId::Submodule).any(|submodule| {
        collect_test_functions(db, submodule).is_empty().not() || has_any_test(db, submodule)
    })
}

fn get_test_lens_position<'db>(db: &'db AnalysisDatabase, ptr: SyntaxStablePtrId<'db>) -> Option<Position> {
    let (original_node, original_file) = get_original_node_and_file(db, ptr)?;
    original_node
        .find_attr(db, "test")
        .map(|attr| attr.as_syntax_node())
        // If attr is not found we are probably on mod.
        .unwrap_or(original_node)
        .span_start_without_trivia(db)
        .position_in_file(db, original_file)
        .map(|position| position.to_lsp())
}

fn maybe_push_code_lens(
    db: &AnalysisDatabase,
    file_state: &mut FileCodeLens,
    make_code_lens: impl FnOnce(Position) -> CodeLens,
    annotated_function: AnnotatedNode,
) {
    let AnnotatedNode { attribute_ptr, full_path: _ } = annotated_function;
    if let Some(position) = get_test_lens_position(db, attribute_ptr) {
        let lens = make_code_lens(position);

        file_state.push(LSCodeLens::Test(TestCodeLens { lens }))
    }
}

fn make_test_code_lens(
    file_url: &Url,
    position: Position,
    is_plural: bool,
    index: usize,
) -> CodeLens {
    let mut title = "▶ Run test".to_string();

    if is_plural {
        title.push('s');
    }

    let command = Command {
        title,
        command: "cairo.executeCodeLens".to_string(),
        arguments: Some(make_lens_args(file_url.clone(), index)),
    };
    let range = Range::new(position, position);

    CodeLens { range, command: Some(command), data: None }
}
