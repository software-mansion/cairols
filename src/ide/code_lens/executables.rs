use crate::ide::code_lens::{
    AnnotatedNode, FileCodeLens, LSCodeLens, LSCodeLensInterface, collect_functions_with_attrs,
    get_original_node_and_file, make_lens_args,
};
use crate::lang::db::AnalysisDatabase;
use crate::lang::lsp::LsProtoGroup;
use crate::lang::lsp::ToLsp;
use crate::lsp::ext::{ExecuteInTerminal, ExecuteInTerminalParams};
use crate::project::builtin_plugins::BuiltinPlugin;
use crate::server::client::Notifier;
use crate::state::State;
use cairo_lang_defs::db::DefsGroup;
use cairo_lang_defs::ids::ModuleId;
use cairo_lang_executable::plugin::EXECUTABLE_ATTR;
use cairo_lang_syntax::node::TypedSyntaxNode;
use cairo_lang_syntax::node::helpers::QueryAttrs;
use cairo_lang_syntax::node::ids::SyntaxStablePtrId;
use cairo_lang_utils::LookupIntern;
use lsp_types::{CodeLens, Command, Position, Range, Url};

#[derive(PartialEq, Clone, Debug)]
pub struct ExecutableCodeLens {
    pub lens: CodeLens,
    command: String,
}

impl LSCodeLensInterface for ExecutableCodeLens {
    fn execute(&self, file_url: Url, state: &State, notifier: &Notifier) -> Option<()> {
        let file_path = file_url.to_file_path().ok()?;
        notifier.notify::<ExecuteInTerminal>(ExecuteInTerminalParams {
            cwd: state.project_controller.configs_registry().manifest_dir_for_file(&file_path)?,
            command: self.command.clone(),
        });
        None
    }

    fn get_lens(&self) -> CodeLens {
        self.lens.clone()
    }
}

pub fn push_executable_code_lenses(
    file_code_lens: &mut FileCodeLens,
    url: Url,
    db: &AnalysisDatabase,
) -> Option<()> {
    let file = db.file_for_url(&url)?;
    let main_module = *db.file_modules(file).ok()?.first()?;

    let crate_id = main_module.owning_crate(db);
    let has_executable_plugin = db
        .crate_macro_plugins(crate_id)
        .iter()
        .filter_map(|plugin_id| {
            BuiltinPlugin::try_from_compiler_macro_plugin(&*plugin_id.lookup_intern(db).0)
                .map(|builtin_plugin: BuiltinPlugin| builtin_plugin == BuiltinPlugin::Executable)
        })
        .next()
        .is_some();

    if !has_executable_plugin {
        return None;
    }

    push_executable_lenses(file_code_lens, db, main_module, url);
    Some(())
}

fn push_executable_lenses(
    file_code_lens: &mut FileCodeLens,
    db: &AnalysisDatabase,
    module: ModuleId,
    file_url: Url,
) {
    for AnnotatedNode { full_path, attribute_ptr } in collect_executable_functions(db, module) {
        if let Some(position) = get_executable_lens_position(db, attribute_ptr) {
            insert_executable_code_lens_pair(file_code_lens, &file_url, &full_path, position);
        }
    }

    let Ok(modules) = db.module_submodules_ids(module) else { return };
    for submodule in modules.iter().copied() {
        if db.is_submodule_inline(submodule) {
            push_executable_lenses(
                file_code_lens,
                db,
                ModuleId::Submodule(submodule),
                file_url.clone(),
            );
        }
    }
}

fn insert_executable_code_lens_pair(
    file_state: &mut FileCodeLens,
    file_url: &Url,
    function_full_path: &String,
    position: Position,
) {
    let range = Range::new(position, position);
    let exec_fn = LSCodeLens::Executable(ExecutableCodeLens {
        lens: CodeLens {
            range,
            command: Some(Command {
                title: String::from("â–¶ Execute function"),
                command: "cairo.executeCodeLens".to_string(),
                arguments: Some(make_lens_args(file_url.clone(), file_state.len())),
            }),
            data: None,
        },
        command: format!("scarb execute --executable-function {function_full_path}"),
    });
    file_state.push(exec_fn);
}

fn collect_executable_functions(db: &AnalysisDatabase, module: ModuleId) -> Vec<AnnotatedNode> {
    collect_functions_with_attrs(db, module, &[EXECUTABLE_ATTR])
}

fn get_executable_lens_position(
    db: &AnalysisDatabase,
    attribute_ptr: SyntaxStablePtrId,
) -> Option<Position> {
    let (original_node, original_file) = get_original_node_and_file(db, attribute_ptr)?;
    original_node.find_attr(db, EXECUTABLE_ATTR).map(|attribute| {
        attribute
            .as_syntax_node()
            .span(db)
            .start
            .position_in_file(db, original_file)
            .map(|position| position.to_lsp())
    })?
}
