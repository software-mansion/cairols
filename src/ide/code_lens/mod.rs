use std::collections::HashMap;
use std::collections::hash_map::Entry;
use std::panic::{AssertUnwindSafe, catch_unwind};
use std::sync::{Arc, RwLock};

use crate::config::Config;
use crate::ide::code_lens::executables::{ExecutableCodeLens, push_executable_code_lenses};
use crate::ide::code_lens::tests::{TestCodeLens, push_test_code_lenses};
use crate::lang::db::{AnalysisDatabase, LsSyntaxGroup};
use crate::server::client::{Notifier, Requester};
use crate::server::schedule::thread::{JoinHandle, ThreadPriority};
use crate::server::schedule::{Task, thread};
use crate::state::State;
use cairo_lang_defs::db::DefsGroup;
use cairo_lang_defs::ids::{ModuleId, ModuleItemId, TopLevelLanguageElementId};
use cairo_lang_filesystem::db::get_originating_location;
use cairo_lang_filesystem::ids::FileId;
use cairo_lang_syntax::node::ast::ModuleItem;
use cairo_lang_syntax::node::helpers::QueryAttrs;
use cairo_lang_syntax::node::ids::SyntaxStablePtrId;
use cairo_lang_syntax::node::{SyntaxNode, TypedStablePtr, TypedSyntaxNode};
use crossbeam::channel::{self, Receiver, Sender};
use itertools::Itertools;
use lsp_types::request::CodeLensRefresh;
use lsp_types::{CodeLens, Url};
use serde::{Deserialize, Serialize};
use serde_json::{Number, Value};

mod executables;
mod tests;

trait LSCodeLensInterface {
    fn execute(&self, file_url: Url, state: &State, notifier: &Notifier) -> Option<()>;
    fn get_lens(&self) -> CodeLens;
}

#[derive(Clone, PartialEq)]
pub enum LSCodeLens {
    Test(TestCodeLens),
    Executable(ExecutableCodeLens),
}

impl LSCodeLens {
    pub fn execute(&self, file_url: Url, state: &State, notifier: &Notifier) -> Option<()> {
        match self {
            LSCodeLens::Test(test_code_lens) => test_code_lens.execute(file_url, state, notifier),
            LSCodeLens::Executable(executable_code_lens) => {
                executable_code_lens.execute(file_url, state, notifier)
            }
        }
    }
    pub fn get_lens(&self) -> CodeLens {
        match self {
            LSCodeLens::Test(test) => test.get_lens(),
            LSCodeLens::Executable(executable) => executable.get_lens(),
        }
    }
}

pub type FileCodeLens = Vec<LSCodeLens>;

#[derive(Default)]
pub struct CodeLensControllerState {
    lens: HashMap<Url, FileCodeLens>,
}

#[derive(Clone)]
pub struct CodeLensController {
    state: Arc<RwLock<CodeLensControllerState>>,
    refresh_sender: Sender<RefreshCodeLensRequest>,
    request_refresh_receiver: Receiver<()>,
    // Keep it last so we can drop channels.
    // Otherwise, the refresh thread will never stop, and the
    // JoinHandle drop will cause deadlock by waiting for the thread to join.
    _refresh_thread: Arc<JoinHandle<()>>,
}

impl CodeLensController {
    pub fn new() -> Self {
        let (refresh_sender, refresh_receiver) = channel::unbounded();
        // If there would be more than single element in queue we should ignore it and send request to client only once.
        // Dedup it on queue level for simplicity.
        let (request_refresh_sender, request_refresh_receiver) = channel::bounded(1);

        let state = Default::default();

        Self {
            state: Arc::clone(&state),
            refresh_sender,
            request_refresh_receiver,
            _refresh_thread: CodeLensRefreshThread::spawn(
                state,
                request_refresh_sender,
                refresh_receiver,
            )
            .into(),
        }
    }

    pub fn request_refresh_receiver(&self) -> Receiver<()> {
        self.request_refresh_receiver.clone()
    }

    pub fn handle_refresh(requester: &mut Requester<'_>) {
        let _ = requester.request::<CodeLensRefresh>((), |_| Task::nothing());
    }

    #[tracing::instrument(skip_all)]
    pub fn schedule_refreshing_all_lenses(&self, db: AnalysisDatabase, config: Config) {
        let lens_guard = self.state.read().unwrap();

        // Invalidate all the files in the state
        let files: Vec<_> = lens_guard
            .lens
            .keys()
            .map(|url| FileChange { url: url.clone(), was_deleted: false })
            .collect();

        // Release so any panickable action is performed while not keeping state lock.
        drop(lens_guard);

        self.schedule_refresh(db, config, files);
    }

    #[tracing::instrument(name = "CodeLensController::on_did_change", skip_all)]
    pub fn on_did_change(
        &self,
        db: AnalysisDatabase,
        config: Config,
        files: impl Iterator<Item = FileChange>,
    ) {
        let lens_guard = self.state.read().unwrap();

        // If it was not requested before, there is nothing to invalidate.
        let files: Vec<_> =
            files.filter(|file_change| lens_guard.lens.contains_key(&file_change.url)).collect();

        // Release so any panickable action is performed while not keeping state lock.
        drop(lens_guard);

        self.schedule_refresh(db, config, files);
    }

    pub fn code_lens(
        &self,
        url: Url,
        db: &AnalysisDatabase,
        config: &Config,
    ) -> Option<Vec<CodeLens>> {
        let lens_state = self.state.read().unwrap();

        let file_code_lens: FileCodeLens = if let Some(code_lens) = lens_state.lens.get(&url) {
            code_lens.clone()
        } else {
            drop(lens_state);

            let result = calculate_code_lens(url.clone(), db, config)?;

            // Lock state only if calculating did *not* panic, so the lock will not be poisoned.
            let mut state = self.state.write().unwrap();
            let entry = state.lens.entry(url);

            entry.insert_entry(result.clone());
            result
        };

        let code_lens = file_code_lens
            .into_iter()
            .map(|lens| lens.get_lens())
            .sorted_by_key(|lens| lens.command.clone().unwrap_or_default().title)
            .collect();

        Some(code_lens)
    }

    pub fn execute_code_lens(state: &State, notifier: Notifier, args: &[Value]) -> Option<()> {
        let (file_url, index) = parse_args(args)?;

        let ls_code_lens = {
            let code_lens_state = state.code_lens_controller.state.read().ok()?;
            let file_lens_state = code_lens_state.lens.get(&file_url)?;
            let item_ref = file_lens_state.get(index)?;
            item_ref.clone()
        };

        ls_code_lens.execute(file_url, state, &notifier);
        Some(())
    }

    #[tracing::instrument(skip_all)]
    fn schedule_refresh(&self, db: AnalysisDatabase, config: Config, files: Vec<FileChange>) {
        let _ = self.refresh_sender.send(RefreshCodeLensRequest { db, config, files });
    }
}

struct RefreshCodeLensRequest {
    db: AnalysisDatabase,
    config: Config,
    files: Vec<FileChange>,
}

struct CodeLensRefreshThread {
    state: Arc<RwLock<CodeLensControllerState>>,
    request_refresh_sender: Sender<()>,
    refresh_receiver: Receiver<RefreshCodeLensRequest>,
}

impl CodeLensRefreshThread {
    fn spawn(
        state: Arc<RwLock<CodeLensControllerState>>,
        request_refresh_sender: Sender<()>,
        refresh_receiver: Receiver<RefreshCodeLensRequest>,
    ) -> JoinHandle<()> {
        let this = Self { state, request_refresh_sender, refresh_receiver };

        thread::Builder::new(ThreadPriority::Worker)
            .name("cairo-ls:code-lens-refresher".into())
            .spawn(move || this.event_loop())
            .expect("failed to spawn code lens refresher thread")
    }

    fn event_loop(self) {
        while let Ok(message) = self.refresh_receiver.recv() {
            let message = self.refresh_receiver.try_iter().fold(message, |mut acc, next_messge| {
                acc.db = next_messge.db; // Leave only single snapshot, drop others.
                acc.config = next_messge.config; // Use last sent config.

                acc.files.extend(next_messge.files);
                acc
            });

            let _ = catch_unwind(AssertUnwindSafe(|| {
                self.refresh_lenses_for(
                    &message.db,
                    &message.config,
                    message.files.into_iter().unique(),
                );
            }));
        }
    }

    #[tracing::instrument(skip_all)]
    fn refresh_lenses_for(
        &self,
        db: &AnalysisDatabase,
        config: &Config,
        files: impl IntoIterator<Item = FileChange>,
    ) {
        // Collect so any panickable action is performed while not keeping state lock.
        let entries: Vec<_> = files
            .into_iter()
            .filter_map(|file_change| {
                calculate_code_lens(file_change.url.clone(), db, config)
                    .map(|code_lenses| (file_change, code_lenses))
            })
            .collect();

        let mut lens_guard = self.state.write().unwrap();

        let mut should_refresh = false;

        for (file_change, code_lenses) in entries {
            let entry = lens_guard.lens.entry(file_change.url.clone());

            should_refresh = should_refresh
                || match &entry {
                    Entry::Occupied(occupied) if file_change.was_deleted => {
                        !occupied.get().is_empty()
                    }
                    Entry::Occupied(occupied) => occupied.get() != &code_lenses,
                    Entry::Vacant(_) => !code_lenses.is_empty(),
                };

            if file_change.was_deleted {
                lens_guard.lens.remove(&file_change.url);
            } else {
                entry.insert_entry(code_lenses);
            }
        }

        if should_refresh {
            let _ = self.request_refresh_sender.try_send(());
        }
    }
}

#[derive(Debug, Clone, PartialEq, Eq, Hash)]
pub struct FileChange {
    pub url: Url,
    pub was_deleted: bool,
}

#[derive(Clone, Copy, PartialEq, Serialize, Deserialize)]
pub enum CodeLensKind {
    Test,
    Executable,
}

fn calculate_code_lens(url: Url, db: &AnalysisDatabase, config: &Config) -> Option<FileCodeLens> {
    let mut result = vec![];

    push_test_code_lenses(&mut result, url.clone(), db, config);
    push_executable_code_lenses(&mut result, url, db);

    Some(result)
}

fn make_lens_args(file_url: Url, lens_index: usize) -> Vec<Value> {
    vec![Value::String(file_url.to_string()), Value::Number(Number::from(lens_index))]
}

fn parse_args(args: &[Value]) -> Option<(Url, usize)> {
    let [Value::String(url), Value::Number(lens_index)] = args else {
        return None;
    };
    let url: Url = url.parse().ok()?;
    let lens_index = lens_index.as_u64().unwrap() as usize;

    Some((url, lens_index))
}

struct AnnotatedNode {
    pub full_path: String,
    pub attribute_ptr: SyntaxStablePtrId,
}
/// Collects functions with given attributes on them
/// Returns tuples of (full path, pointer to found attribute)
fn collect_functions_with_attrs(
    db: &AnalysisDatabase,
    module: ModuleId,
    attributes: &[&str],
) -> Vec<AnnotatedNode> {
    let mut result = vec![];

    if let Ok(functions) = db.module_free_functions(module) {
        for (free_function_id, function) in functions.iter() {
            let function_full_path = ModuleItemId::FreeFunction(*free_function_id).full_path(db);
            result.extend(
                attributes
                    .iter()
                    .filter_map(|attr_name| function.find_attr(db, attr_name))
                    .map(|attr| AnnotatedNode {
                        full_path: function_full_path.clone(),
                        attribute_ptr: attr.stable_ptr(db).untyped(),
                    })
                    // If for some reason we found multiple attributes relevant for the code lens kind, push only the first one.
                    .next(),
            );
        }
    }

    result
}

fn get_original_node_and_file(
    db: &AnalysisDatabase,
    ptr: SyntaxStablePtrId,
) -> Option<(SyntaxNode, FileId)> {
    let (file, span) =
        get_originating_location(db, ptr.file_id(db), ptr.lookup(db).span_without_trivia(db), None);

    db.find_syntax_node_at_offset(file, span.start)?
        .ancestors_with_self(db)
        .find(|n| ModuleItem::cast(db, *n).is_some())
        .map(|syntax_node| (syntax_node, file))
}
