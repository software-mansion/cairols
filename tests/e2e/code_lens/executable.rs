use crate::code_lens::test_code_lens_scarb_execute;
use indoc::indoc;

#[test]
fn test_one_executable_whole_package() {
    insta::assert_snapshot!(test_code_lens_scarb_execute((
        r#"
        #[executable]<caret>
        fn main(){
            // Do stuff
        }
        "#,
        indoc!(
            r#"
            [package]
            name = "indor"
            version = "0.1.0"
            edition = "2024_07"

            [dependencies]
            cairo_execute = "2.9.0"

            [executable]
            "#
        )
    )));
}

#[test]
fn test_two_executables_whole_package() {
    insta::assert_snapshot!(test_code_lens_scarb_execute((
        r#"
        #[executable]<caret>
        fn main(){
            // Do stuff
        }
        #[executable]
        fn main2(){
            // Do stuff
        }
        "#,
        indoc!(
            r#"
            [package]
            name = "indor"
            version = "0.1.0"
            edition = "2024_07"

            [dependencies]
            cairo_execute = "2.9.0"

            [executable]
            "#
        )
    )));
}

#[test]
fn test_two_executables_one_precise_def() {
    insta::assert_snapshot!(test_code_lens_scarb_execute((
        r#"
        #[executable]<caret>
        fn main(){
            // Do stuff
        }
        #[executable]
        fn main2(){
            // Do stuff
        }
        "#,
        indoc!(
            r#"
            [package]
            name = "indor"
            version = "0.1.0"
            edition = "2024_07"

            [dependencies]
            cairo_execute = "2.9.0"

            [[target.executable]]
            function = "indor::main2"
            "#
        )
    )));
}
