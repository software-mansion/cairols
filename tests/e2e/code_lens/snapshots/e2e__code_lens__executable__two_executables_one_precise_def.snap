---
source: tests/e2e/code_lens/executable.rs
expression: "test_code_lens_scarb_execute((r#\"\n        #[executable]<caret>\n        fn main(){\n            // Do stuff\n        }\n        #[executable]\n        fn main2(){\n            // Do stuff\n        }\n        \"#,\nindoc!(r#\"\n            [package]\n            name = \"indor\"\n            version = \"0.1.0\"\n            edition = \"2024_07\"\n\n            [dependencies]\n            cairo_execute = \"2.9.0\"\n\n            [[target.executable]]\n            function = \"indor::main2\"\n            \"#)))"
---
[[lenses]]
line = 0
command = "▶ Execute function"
file_path = "src/lib.cairo"
index = 0

[[lenses]]
line = 4
command = "▶ Execute function"
file_path = "src/lib.cairo"
index = 1

[execute_in_terminal]
command = "scarb execute --executable-function indor::main"
cwd = "./"
